File: lizzie-mobile/shared/src/commonMain/kotlin/com/lizzie/rules/BoardData.kt
package com.lizzie.rules

import com.lizzie.analysis.MoveData

/**
 * Holds the full state of a single board position, including stones,
 * analysis data, and move metadata.
 */
data class BoardData(
    val stones: StoneArray,
    val lastMove: Pair<Int, Int>?,
    val lastMoveColor: Stone,
    val blackToPlay: Boolean,
    val zobrist: Zobrist = Zobrist(),
    val moveNumber: Int = 0,
    val moveMNNumber: Int = -1,
    val moveNumberList: IntArray,
    val blackCaptures: Int = 0,
    val whiteCaptures: Int = 0,
    val winrate: Double = 50.0,
    val playouts: Int = 0,
    val scoreMean: Double = 0.0,
    val engineIndex: Int = -1,
    val komi: Double = 6.5,
    val bestMoves: List<MoveData> = emptyList(),
    val comment: String = "",
    val properties: Map<String, String> = emptyMap(),
    val dummy: Boolean = false,
) {
    fun withBestMoves(moves: List<MoveData>): BoardData {
        val p = MoveData.getPlayouts(moves)
        if (p > playouts) {
            return copy(
                bestMoves = moves,
                playouts = p,
                winrate = MoveData.getWinrateFromBestMoves(moves),
                scoreMean = MoveData.getScoreMeanFromBestMoves(moves)
            )
        }
        return this
    }

    fun withClearedBestMoves(engineIndex: Int, komi: Double): BoardData {
        return copy(
            bestMoves = emptyList(),
            playouts = 0,
            engineIndex = engineIndex,
            komi = komi,
        )
    }

    fun getDisplayWinrate(alwaysBlackWinrate: Boolean): Double {
        return if (!blackToPlay || !alwaysBlackWinrate) {
            winrate
        } else {
            100.0 - winrate
        }
    }

    fun getDisplayScoreMean(): Double {
        return if (blackToPlay) scoreMean else -scoreMean
    }

    fun isSameCoord(coord: Pair<Int, Int>?): Boolean {
        if (coord == null || lastMove == null) return false
        return lastMove == coord
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is BoardData) return false
        return zobrist == other.zobrist &&
                comment == other.comment &&
                blackToPlay == other.blackToPlay
    }

    override fun hashCode(): Int = zobrist.hashCode()

    companion object {
        fun empty(width: Int = 19, height: Int = 19): BoardData {
            val size = width * height
            return BoardData(
                stones = StoneArray.empty(size),
                lastMove = null,
                lastMoveColor = Stone.EMPTY,
                blackToPlay = true,
                zobrist = Zobrist(),
                moveNumber = 0,
                moveNumberList = IntArray(size),
                blackCaptures = 0,
                whiteCaptures = 0,
                winrate = 50.0,
                playouts = 0,
                scoreMean = 0.0,
            )
        }
    }
}

File: lizzie_mobile/lib/go/board_data.dart
import 'stone.dart';

/// Immutable snapshot of the board at one node in the move history tree.
///
/// Ported from Lizzie's `BoardData.java`.
class BoardData {
  final int width;
  final int height;

  /// Column-major flat array: index = x * height + y.
  final List<Stone> stones;

  /// [x, y] of last move, or null for pass / root.
  final List<int>? lastMove;

  /// Color of the last move played.
  final Stone lastMoveColor;

  /// True if it's Black's turn now.
  final bool blackToPlay;

  /// Incremental Zobrist hash of this position.
  final int zobrist;

  /// Move number (1-based).
  final int moveNumber;

  /// Per-cell display move number (0 = empty, otherwise move number).
  final List<int> moveNumberList;

  /// Stones captured BY black (i.e. black captured this many white stones).
  final int blackCaptures;

  /// Stones captured BY white.
  final int whiteCaptures;

  /// Current winrate (0-100, side-to-move perspective).
  final double winrate;

  /// Current total playouts across all suggested moves.
  final int playouts;

  /// Score mean from KataGo analysis.
  final double scoreMean;

  /// SGF MN property override (-1 means not set).
  final int moveMNNumber;

  /// Marker node (e.g. SGF AB[] passthrough placeholder).
  final bool dummy;

  /// SGF node properties (comments, labels, etc.).
  final Map<String, String> properties;

  /// Comment text for this node.
  String comment;

  BoardData({
    required this.width,
    required this.height,
    required this.stones,
    this.lastMove,
    required this.lastMoveColor,
    required this.blackToPlay,
    required this.zobrist,
    required this.moveNumber,
    required this.moveNumberList,
    this.blackCaptures = 0,
    this.whiteCaptures = 0,
    this.winrate = 50.0,
    this.playouts = 0,
    this.scoreMean = 0.0,
    this.moveMNNumber = -1,
    this.dummy = false,
    Map<String, String>? properties,
    this.comment = '',
  }) : properties = properties ?? {};

  factory BoardData.empty({int width = 19, int height = 19}) {
    final n = width * height;
    return BoardData(
      width: width,
      height: height,
      stones: List<Stone>.filled(n, Stone.empty),
      lastMove: null,
      lastMoveColor: Stone.empty,
      blackToPlay: true,
      zobrist: 0,
      moveNumber: 0,
      moveNumberList: List<int>.filled(n, 0),
    );
  }

  /// The number of intersections.
  int get size => width * height;

  int indexOf(int x, int y) => x * height + y;

  bool isValid(int x, int y) =>
      x >= 0 && x < width && y >= 0 && y < height;
}
